core/system/src/page_allocator.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This file can be compiled externally to provide the page allocator algorithm
+   for devices NOT supported by uVisor. For this purpose this file is copied as
+   is into uvisor-mbed-lib and compiled by the target build system. */
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
+#include <uvisor.h>
+#include "page_allocator.h"
+#include "api/inc/vmpu_unpriv_access_exports.h"
+#include "mpu/vmpu.h"
+#include "halt.h"
+#include "context.h"
+#else
+#include "uvisor-lib/uvisor-lib.h"
+#define DPRINTF(...) {}
+#define g_active_box 0
+#define vmpu_is_box_id_valid(...) 0
+#define vmpu_public_flash_addr(...) 1
+#define vmpu_sram_addr(...) 1
+#define HALT_ERROR(id, ...) DPRINTF(__VA_ARGS__)
+static int vmpu_is_region_size_valid(uint32_t size)
+{
+    /* Minimal 32B alignment */
+    return ((size & ~31) == size);
+}
+static uint32_t vmpu_round_up_region(uint32_t addr, uint32_t size)
+{
+    if (!vmpu_is_region_size_valid(size)) {
+        return 0;
+    }
+    const uint32_t mask = size - 1;
+    return (addr + mask) & ~mask;
+}
+#endif
+
+/* By default a maximum of 16 pages are allowed. This can only
+   be decided by the porting engineer for the current platform! */
+#ifndef UVISOR_PAGE_TABLE_MAX_COUNT
+#define UVISOR_PAGE_TABLE_MAX_COUNT ((size_t) 16)
+#endif
+/* The page box_id is the box id which is 8bit large  */
+typedef uint8_t page_owner_t;
+/* Maps the page to the owning box handle */
+static page_owner_t g_page_owner_table[UVISOR_PAGE_TABLE_MAX_COUNT];
+/* define a unused value for the page table */
+#define UVISOR_PAGE_UNUSED ((page_owner_t) (-1))
+
+/* contains the configured page size */
+static uint32_t g_page_size;
+/* Points to the beginning of the page heap */
+static const void * g_page_heap_start;
+/* Points to the end of the page heap */
+static const void * g_page_heap_end;
+/* Contains the number of free pages */
+static uint8_t g_page_count_free;
+/* Contains the total number of available pages */
+static uint8_t g_page_count_total;
+
+
+void page_allocator_init(void * const heap_start, void * const heap_end, uint32_t * const page_size)
```

Comment:
I guess page_size is a `uint32_t const * const` (constant pointer to constant uint32)?